apps/prairielearn/src/tests/render.test.ts

```diff
@@ -0,0 +1,454 @@
+import { dirname, join, resolve } from 'path';
+import { fileURLToPath } from 'url';
+
+import { config as chaiConfig } from 'chai';
+import { assert } from 'chai'; // Import assert
+// import chai from 'chai';
+import fs from 'fs-extra';
+import { HTMLRewriter } from 'html-rewriter-wasm';
+import { HtmlValidate } from 'html-validate';
+import { HTMLHint } from 'htmlhint';
+import * as log from 'why-is-node-running';
+
+import * as assetServer from '../lib/assets.js';
+import * as codeCaller from '../lib/code-caller/index.js';
+import { config } from '../lib/config.js';
+import type { Course, Question, Submission, Variant } from '../lib/db-types.js';
+import { features } from '../lib/features/index.js';
+import * as loadEstimator from '../lib/load.js';
+import { buildQuestionUrls } from '../lib/question-render.js';
+import { makeVariant } from '../lib/question-variant.js';
+import * as freeformServer from '../question-servers/freeform.js';
+import * as questionServers from '../question-servers/index.js';
+import * as helperServer from '../tests/helperServer.js';
+const htmlvalidate = new HtmlValidate();
+
+const __filename = fileURLToPath(import.meta.url);
+const __dirname = dirname(__filename);
+
+const exampleCoursePath = resolve(__dirname, '..', '..', '..', '..', 'exampleCourse');
+const questionsPath = join(exampleCoursePath, 'questions');
+chaiConfig.showDiff = true;
+
+const buildSubmission = (answer: any): any => {
```

Comment:
This is an adhoc way to recover the original input. Other than fixing up the types, are we fine going with a "if it works" mentality for this test?

Reply:
Have you thought of using the question-testing.ts library and exporting [`createTestSubmission`](https://github.com/PrairieLearn/PrairieLearn/blob/b1929d0341d01e6ed1d94bfb3e1293a89a435d9a/apps/prairielearn/src/lib/question-testing.ts#L54)? This would create a submission based on input values provided by elements themselves.

Reply:
Good idea. I can build the submission through this.